nevergrad/functions/powersystems/core.py

```diff
@@ -0,0 +1,233 @@
+# University Clermont Auvergne, CNRS, SIGMA Clermont, Institut Pascal
+
+import copy
+from math import sqrt, tan, pi
+import math
+import matplotlib.pyplot as plt
+from typing import Any
+from typing import List
+import numpy as np
+from nevergrad.common.typetools import ArrayLike
+from ... import instrumentation as inst
+from ...instrumentation.core import Variable
+from ...instrumentation.multivariables import Instrumentation
+pi = np.pi
+
+
+class Agent():
+    """An agent has an input size, an output size, a number of layers, a width of its internal layers 
+    (a.k.a number of neurons per hidden layer)."""
+
+    def __init__(self, input_size: int, output_size: int, layers: int=3, layer_width: int=14):
+        assert(layers >= 2)
+        self.input_size = input_size
+        self.output_size = output_size
+        self.layers: List[Any] = []
+        self.layers += [np.zeros((input_size, layer_width))]
+        for i in range(layers-2):
+            self.layers += [np.zeros((layer_width, layer_width))]
+        self.layers += [np.zeros((layer_width, output_size))]
+        assert len(self.layers) == layers
+
+    def GetParamNumbers(self):
+        return sum([np.prod(l.shape) for l in self.layers])
+
+    def SetParams(self, ww):
+        w = [w for w in ww]
+        assert(len(w) == self.GetParamNumbers())
+        for l in self.layers:
+            s = np.prod(l.shape)
+            l = np.reshape(np.array(w[:s]), l.shape)
+            w = w[s:]
+
+    def GetOutput(self, inp):
+        output = np.array(inp).reshape(1, len(inp))
+        for l in [self.layers[i] for i in range(len(self.layers) - 1)]:
+            output = np.tanh(np.matmul(output, l))
+        return np.matmul(output, self.layers[-1])
+
+
+# Real life is more complicated! This is a very simple model.
+class PowerSystem(inst.InstrumentedFunction):
+    """
+    Parameters
+    ----------
+    nint intaum_stocks: number of stocks to be managed
+    depth: number of layers in the neural networks
+    width: number of neurons per hidden layer
+    """
+
+    def __init__(self, num_stocks: int = 13, depth: int = 3, width: int = 3, 
+            year_to_day_ratio: float = 2.,  # Ratio between std of consumption in the year and std of consumption in the day.
+            constant_to_year_ratio: float = 1.,  # Ratio between constant baseline consumption and std of consumption in the year.
+            back_to_normal: float = 0.5,  # Part of the variability which is forgotten at each time step.
+            consumption_noise: float = 0.1,  # Instantaneous variability.
+            num_thermal_plants: int = 7,  # Number of thermal plants.
+            num_years: int = 1,  # Number of years.
+            failure_cost: float = 500.,  # Cost of not satisfying the demand. Equivalent to an expensive infinite capacity thermal plant.
+            ) -> None:
+        # Number of stocks (dams).
+        self.N = num_stocks
```

Comment:
Call it `self.num_stocks` not `N`. Actually `num_dams` would probably be clearer (you wouldn't need a comment to understand what it is).

Reply:
Done.